bin/metrics-alb.rb

```diff
@@ -0,0 +1,162 @@
+#
+# LICENSE:
+#   Copyright 2018 Jonathan Ballet <jon@multani.info>
+#   Released under the same terms as Sensu (the MIT license); see LICENSE
+#   for details.
+
+require 'sensu-plugin/metric/cli'
+require 'aws-sdk'
+require 'sensu-plugins-aws'
+require 'time'
+
+class ALBMetrics < Sensu::Plugin::Metric::CLI::Graphite
+  include CloudwatchCommon
+
+  option :albname,
+         description: 'Name of the Application Load Balancer',
+         short: '-n ALB_NAME',
+         long: '--name ALB_NAME'
+
+  option :scheme,
+         description: 'Metric naming scheme, text to prepend to metric',
+         short: '-s SCHEME',
+         long: '--scheme SCHEME',
+         default: 'alb'
+
+  option :fetch_age,
+         description: 'How long ago to fetch metrics for',
+         short: '-f AGE',
+         long: '--fetch-age',
+         default: 60,
+         proc: proc(&:to_i)
+
+  option :aws_region,
+         short: '-r AWS_REGION',
+         long: '--aws-region REGION',
+         description: 'AWS Region (defaults to the AWS_REGION variable)',
+         default: ENV['AWS_REGION']
+
+  option :end_time,
+         short:       '-t TIME',
+         long:        '--end-time TIME',
+         default:     Time.now,
+         proc:        proc { |a| Time.parse a },
+         description: 'CloudWatch metric statistics end time'
+
+  option :period,
+         short:       '-p SECONDS',
+         long:        '--period SECONDS',
+         default:     60,
+         proc:        proc(&:to_i),
+         description: 'CloudWatch metric statistics period'
+
+  def loadbalancer
+    @loadbalancer = Aws::ElasticLoadBalancingV2::Client.new
+  end
+
+  def cloud_watch_metric(metric_name, value, load_balancer_name, alb_id)
+    name = ['app', load_balancer_name, alb_id].join('/')
+    client.get_metric_statistics(
+      namespace: 'AWS/ApplicationELB',
+      metric_name: metric_name,
+      dimensions: [
+        {
+          name: 'LoadBalancer',
+          value: name
+        }
+      ],
+      statistics: [value],
+      start_time: config[:end_time] - config[:fetch_age] - config[:period],
+      end_time: config[:end_time] - config[:fetch_age],
+      period: config[:period]
+    )
+  end
+
+  def print_statistics(load_balancer_name, alb_id, statistics)
+    result = {}
+    static_value = {}
+    statistics.each do |key, static|
+      r = cloud_watch_metric(key, static, load_balancer_name, alb_id)
+      keys = if config[:scheme] == ''
+               []
+             else
+               [config[:scheme]]
+             end
+      keys.concat [load_balancer_name, key, static]
+      metric_key = keys.join('.')
+
+      static_value[metric_key] = static
+      result[metric_key] = r[:datapoints][0] unless r[:datapoints][0].nil?
+    end
+    result.each do |key, value|
+      output key, value[static_value[key].downcase], value[:timestamp].to_i
+    end
+  end
+
+  def run
+    statistics = {
+      'ActiveConnectionCount'          => 'Sum',
```

Comment:
> so looking at the functions I am not quite sure I understand why we need this, can you be specific as to what you don't like about the existing methods?

So, the equivalent function in the `CloudWatchCommon` doesn't support "fetch age", like I added [here](https://github.com/sensu-plugins/sensu-plugins-aws/pull/286) for example. 
Instead, it tries to do something a bit similar by increase the priod time (and I'm not sure if that's better? It's at best slightly different)
The end time config support is OK for testing purpose but I don't see how it could be used in an automated way since this command line argument would have to be changed for every calls. 

"Fetch age" proposes a way to solve that, in that case it could be moved directly into the common library and all the other metrics checks which are going to use the library could take benefit from it without having to reimplement it.

Would that be OK with you?

> Also I'd prefer cloudwatch over cloud_watch as its not watching any cloud its a product called CloudWatch.

I have no problem with that :)

(Actually, the balance between 1. keeping things organized like the other checks/metrics or 2. improve small things like and parting away for existing checks/metrics is difficult to find :( )

> should we param these? We could have sensible defaults but I'd like it to be more flexible as for example I tend to care more about percentiles than average.

That's a good question. Ideally, if it's possible to parameterized these, both the metric name and the type of statistic should be parameterized.

Currently, metrics checks are implemented that kind of selection all in different ways:

* `metrics-asg.rb` allows to change the statistic, but not the metric name
* `metrics-waf.rb` allows to filter the list of metrics/statistics to collect only one of the predefined ones instead of all
* `metrics-elb.rb` statistic selection is actually broken (it replaces a hash by a string)
* and actually, none of them allow to get several statistics for the same metric

We could add a new type of parameter which could help with that, like `--metric-statistic TargetResponseTime=Minimum,Maximum,Average,p99.00`. We could parse this and add/override this from the default metrics collected. The statistics associated to a metric would be then a list instead of the current string. I don't know if it should be in this PR though.

Reply:
Sorry for the delay I missed the notification. Gotcha, we should probably look at refactoring the other approach to use a single method for consistency at a later date. IMO only blocker is renaming of cloudwatch to be consistent.